### PR TITLE
feat: codemirror migration

### LINKED ISSUE #1737: Migrate `editor` to CodeMirror
This is a high-level issue about migrating the editor component from monaco to CodeMirror. Some motivations (including those from [repl.it](https://blog.replit.com/codemirror)):

* Mobile support
* Plugins such as [codemirror-interact](https://github.com/replit/codemirror-interact)
* Good excuse to work on LSP as well

### PR DESCRIPTION
**Pull request by KyleleeSea**
# Description
**not ready to merge yet**
Resolves #1737 #1099.
Replaces Monaco with Codemirrror for the benefits of semantic syntax highlighting, smarter autocomplete, and mobile support. 

# Features
### Semantic Syntax Highlighting
The previous Monaco system utilized regex to highlight which naively colored words certain colors no matter where they appeared in the program. This PR colors based on syntax tree context. 

### Smarter Autocomplete...
**Domain:**
- Keyword autocompletion (smarter than Monaco version, only suggested where appropriate) 
- Type name autocompletion inside of param lists, subtype inheritance, and function and constructor output 

**Substance:**
- Keyword, Type name, and Predicate autocompletion at beginning of statement
- Suggestion of "for", "where", "All" only where appropriate
- Identifier caching and suggestion inside param lists
- Function and constructor name suggestion after assignment operator

**Style:**
- Shape specific property autocomplete
- Caching of namespaces and properties and suggestion inside expressions
- Collector and selector header completion (where, with, foreach)
- Top level keyword suggestion (forall, collect, layout)
- Expression autocomplete 
- Constraints suggested after encourage and ensure
- Block body statement completion (type names, kws, anonymous shapes)
- Type names (user defined from domain) suggested in headers where appropriate
- Repeatable suggested after first header keyword for selectors and collectors
- "into" suggested in collector header
- Predicates suggested after "where" in header 

### Linting / Inline errors
Previously inline errors were only shown after the user clicked compile. Now before compile time, parse errors are shown inline. When the user clicks compile, compiler errors are shown inline. Once the user starts editing again, the compiler errors are hidden and parse errors are shown. 

### Languages Caches
Previously information was shared from domain into substance using "DomainCache" which was constructed by running a compilation of domain and passing an object from the found types. This approach does not work for this use case since we'd like to use this information in domain, but compiling domain while the user is editing will return an error. Instead we traverse the parse tree to gather this information. We currently maintain the following information:
Domain: All function, constructor, predicate, and type names
Substance: All identifier names
Style: All namespace names and properties (except canvas)

### History Clear
Solves issue #1786.
Previously there was no history state reset between changing workspaces, so cmd+z on a newly loaded workspace would reset its content to the previously open workspace.
This is solved in this PR albeit with a hacky solution. Please see: https://github.com/uiwjs/react-codemirror/issues/405
The official recommended way to reset history is to create a new Editor State (https://discuss.codemirror.net/t/codemirror-6-cm-clearhistory-equivalent/2851). This is problematic as this would remove the state set by React Codemirror. 
We use React Codemirror because it provides a non trivial wrapper for Codemirror that abstracts away issues that would take a decent deal of work to deal with. Namely, experimenting with manually using Codemirror in React created the following issues: 1) Need to manually implement the onChange functionality React Codemirror gives us for free 2) Type err on the div 3) Line numbers wouldn't show up.
We also tried using a [compartment](https://codemirror.net/examples/config/) extension however we could not get this to update the editor state, calling .reconfigure on its own didn't make any update, which makes me believe we'd need to trigger a transaction. This would require getting access to the editor State inside React Codemirror, but I couldn't find a way to extract this. Additionally, I'm unsure if resetting by passing a new history() object actually resets the history state (it may just reset the config).

### Test Suite
Comprehensive test suite for all languages in the trio. The following aspects are tested:
- Parser to check that valid programs have no error nodes and invalid programs contain error nodes
- "Caching" to check that the DomainCache contains appropriate type names, predicate names, function names, constructor names. Checks that namespace object in style is correct. Checks that substance id cache is correct.
- Autocomplete. Simulate autocomplete context and checks an expected list of completion items against the returned completion items.

Some notes on the autocomplete testing suite...
- A separate codemirror function narrows down the options based on the current node. This is not tested by our testing suite. We just test whether the correct options appear for the general situation.
- For some style checks we check that our expected list is a subset of the returned options. We do this since some completions (like for expressions) have a substantial number of options and if we ever change this list, we want the test to be robust to future changes.
- Most style autocomplete tests are small, it may be worthwhile in the future to include larger examples

### Vim Mode
Uses: [replit/codemirror-vim](https://github.com/replit/codemirror-vim)
I've tested this turns on, but haven't tested Vim functionality in depth since I don't know Vim.

### Misc
On mobile devices, autocomplete info boxes are hidden as they take up too much space
Edgeworth uses monaco, I left around some monaco functions that are used by Edgeworth 

# Checklist
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have reviewed any generated registry diagram changes

# Known issues
1. Color picker...
A. On mobile, the color "slider" does not function as a slider and closes a second or so after the user taps on the slider
B. If a color literal token is touching another token, selection in the color picker UI has no effect
To reproduce: "a = setOpacity(#abcdef)", click on #abcdef and select a color, observe no change in the editor
Interestingly when you do this inside inspect with the mobile device emulation on, the color picker works as one would expect
C. Color picker has no alpha channel 

2. In domain if a user types a name for the return of a function or constructor, top level keywords will be suggested
Example: "function A() -> B name"
This is because of ambiguity of the following identifier. We make this as an intentional decision so that keywords will be suggested in the following line if user does not name the return object. This is due to the decision to have the parser ignore new lines. 

# Open Questions 
1. Should inline syntax error messages be more descriptive?
Error nodes in the tree do not come with built-in error messages, the only information provided is that the node isError. 
To create more descriptive syntax errors we'd have to traverse the tree and create an error message generator function manually. 

# Future Directions
1. Hover tool tips on shapes for discovery
Example: Hovering on "Circle" will show the properties that it has

2. Block folding
Drop down arrows in the gutter that will fold blocks 

3. Property suggestion filtering
Inside shapes, if a property is already defined, do not suggest it again

4. More descriptive info boxes
Could provide usage examples
Could provide more parameter information 

5. Replace the nearley parser entirely with the Lezer parser defined here

### CODE REVIEW

File: packages/components/src/editing/hooks/domain/getDomainCache.ts [hunk 1]
```diff
@@ -0,0 +1,66 @@
...
+    }
+  });
+
+  return types;
+};
+
+const getPredicateNames = (domainProg: string, predNodes: SyntaxNode[]) => {
+  let preds = [] as string[];
+
+  predNodes.forEach((node: SyntaxNode) => {
```
Comment (wodeni) [inline packages/components/src/editing/hooks/domain/getDomainCache.ts:26]: I think you can just do `const id = node.getChild("Identifier")` to get the first child too?

**Review by wodeni — CHANGES_REQUESTED** [review]
Still need to review a few files in `editor`. Smaller things noted in inline comments. Here're a few high level things I did on #1810:

* The `Listing` component (used in `edgeworth` so far) still uses monaco. Refactored it to be a thin wrapper over `EditorPane`.
* `storybook` somehow throws [this error](https://github.com/WebAssembly/spec/issues/1031). I ended up bumping to storybook v8 because it has slightly better vite and ts support, but the real issue was we did `exclude: ["rose"]` in the vite config. 
  * One more side quest: turns out we were using `18.0` React types, and I bumped `@types/react` and `@types/react-dom` in the rest of the repo because the sb8 migration automatically bumped them. 
* Remove recoil stuff from `EditorPane` because `components` need to be pure react.

File: packages/components/src/editing/hooks/substance/substanceAutocomplete.ts [hunk 2]
```diff
@@ -0,0 +1,180 @@
...
+    let nodeBefore = syntaxTree(context.state).resolveInner(context.pos, -1);
+    let parentNode = nodeBefore.parent;
+    let leftSib = nodeBefore.prevSibling;
+    // In erorr state, error node wraps the current node
+    if (parentNode != null && parentNode.type.isError) {
+      leftSib = parentNode.prevSibling;
+      parentNode = parentNode.parent;
+    }
+
+    let word = context.matchBefore(/\w*/);
```
[the remark this answers is not in the document]
Comment (KyleleeSea) [reply packages/components/src/editing/hooks/substance/substanceAutocomplete.ts:92]: Do you mean a wrapper for just line 92 like:
typingWord(context) return context.matchBefore(/\w*/);

Or something different?

File: packages/components/src/editing/hooks/domain/getDomainCache.ts [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+  });
+
+  return preds;
+};
+
+const getFn_ConsNames = (domainProg: string, nodes: SyntaxNode[]) => {
+  let names = [] as string[];
+
+  nodes.forEach((node: SyntaxNode) => {
+    let nodeCursor = node.cursor();
```
[the remark this answers is not in the document]
Comment (KyleleeSea) [reply packages/components/src/editing/hooks/domain/getDomainCache.ts:45]: Implemented these changes, now the code for get type names, get pred names, and get fn cons names is all the same so I condensed into one getNames function

File: packages/components/src/editing/hooks/domain/getDomainCache.ts (hunk 1 above)
Comment (KyleleeSea) [reply packages/components/src/editing/hooks/domain/getDomainCache.ts:26]: Good catch, forgot about this trick. Thanks!

File: packages/components/src/editing/hooks/substance/getSubstanceCache.ts [hunk 4]
```diff
@@ -0,0 +1,71 @@
...
+  typeAppNodes.forEach((node: SyntaxNode) => {
+    // Ignore indexed statements
+    if (containsIndexedStatement(node)) {
+      return;
+    }
+
+    let nodeCursor = node.cursor();
+    if (!traverseCursorDown(nodeCursor, "NamedId")) return;
+    // move to identifier, loop because of comma sep list
+    while (nodeCursor.nextSibling()) {
```
[the remark this answers is not in the document]
Comment (KyleleeSea) [reply packages/components/src/editing/hooks/substance/getSubstanceCache.ts:21]: Yes, this works

**Review by wodeni — APPROVED** [review]

**Pull request merged**